### PR TITLE
Bump llvm version to llvm-18

### PR DESCRIPTION
**Pull request by hsharsha**

### CODE REVIEW

**Review by i-chaochen — APPROVED** [review]

**Comment by hsharsha** [comment]
I am getting this linker error with llvm-18. @jayfurmanek @i-chaochen  any idea why?
```
ERROR: /home/hahavanu/tensorflow-upstream/tensorflow/BUILD:1316:21: Linking tensorflow/libtensorflow_cc.so.2.18.0 failed: (Exit 1): crosstool_wrapper_driver_is_not_gcc failed: error executing command (from target //tensorflow:libtensorflow_cc.so.2.18.0) external/local_config_rocm/crosstool/clang/bin/crosstool_wrapper_driver_is_not_gcc @bazel-out/k8-opt/bin/tensorflow/libtensorflow_cc.so.2.18.0-2.params
/root/.cache/bazel/_bazel_root/7ff435f49cc74c1ec3639370c9fccbbe/execroot/org_tensorflow/external/local_config_rocm/crosstool/clang/bin/crosstool_wrapper_driver_is_not_gcc:168: SyntaxWarning: invalid escape sequence '\.'
  re.search('\.cpp$|\.cc$|\.c$|\.cxx$|\.C$', f)]
/root/.cache/bazel/_bazel_root/7ff435f49cc74c1ec3639370c9fccbbe/execroot/org_tensorflow/external/local_config_rocm/crosstool/clang/bin/crosstool_wrapper_driver_is_not_gcc:23: DeprecationWarning: 'pipes' is deprecated and slated for removal in Python 3.13
  import pipes
ld.lld: error: undefined symbol: absl::lts_20230802::log_internal::LogMessage& absl::lts_20230802::log_internal::LogMessage::operator<<<int, 0>(int const&)
>>> referenced by cholesky_op_gpu.cu.cc
>>>               bazel-out/k8-opt/bin/tensorflow/core/kernels/linalg/_objs/cholesky_op_gpu/cholesky_op_gpu.cu.pic.o:(tensorflow::CholeskyOpGpu<float>::ComputeAsync(tensorflow::OpKernelContext*, std::function<void ()>)::'lambda'(absl::lts_20230802::Status const&, std::vector<tensorflow::HostLapackInfo, std::allocator<tensorflow::HostLapackInfo>> const&)::operator()(absl::lts_20230802::Status const&, std::vector<tensorflow::HostLapackInfo, std::allocator<tensorflow::HostLapackInfo>> const&) const)
>>> referenced by cholesky_op_gpu.cu.cc
>>>               bazel-out/k8-opt/bin/tensorflow/core/kernels/linalg/_objs/cholesky_op_gpu/cholesky_op_gpu.cu.pic.o:(tensorflow::CholeskyOpGpu<double>::ComputeAsync(tensorflow::OpKernelContext*, std::function<void ()>)::'lambda'(absl::lts_20230802::Status const&, std::vector<tensorflow::HostLapackInfo, std::allocator<tensorflow::HostLapackInfo>> const&)::operator()(absl::lts_20230802::Status const&, std::vector<tensorflow::HostLapackInfo, std::allocator<tensorflow::HostLapackInfo>> const&) const)
>>> referenced by cholesky_op_gpu.cu.cc
>>>               bazel-out/k8-opt/bin/tensorflow/core/kernels/linalg/_objs/cholesky_op_gpu/cholesky_op_gpu.cu.pic.o:(tensorflow::CholeskyOpGpu<std::complex<float>>::ComputeAsync(tensorflow::OpKernelContext*, std::function<void ()>)::'lambda'(absl::lts_20230802::Status const&, std::vector<tensorflow::HostLapackInfo, std::allocator<tensorflow::HostLapackInfo>> const&)::operator()(absl::lts_20230802::Status const&, std::vector<tensorflow::HostLapackInfo, std::allocator<tensorflow::HostLapackInfo>> const&) const)
>>> referenced 1 more times
clang: error: linker command failed with exit code 1 (use -v to see invocation)
Target //tensorflow/compiler/tests:adagrad_da_test_gpu failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 5060.833s, Critical Path: 589.13s
INFO: 20752 processes: 1762 internal, 18990 local.
FAILED: Build did NOT complete successfully
```

**Comment by i-chaochen** [comment]
I remember @zoranjovanovic-ns just fixed a llvm 18 link issue by https://github.com/ROCm/tensorflow-upstream/commit/cfeeae98c4289fb7043b50edfe39221f726c5f36

**Comment by hsharsha** [comment]
> root/.cache/bazel/_bazel_root/7ff435f49cc74c1ec3639370c9fccbbe/execroot/org_tensorflow/external/local_config_rocm/crosstool/clang/bin/crosstool_wrapper_driver_is_not_gcc:168: SyntaxWarning: invalid escape sequence '\.'

That commit is not part of weekly-sync But it is giving error in escape sequence in re.search in
root/.cache/bazel/_bazel_root/7ff435f49cc74c1ec3639370c9fccbbe/execroot/org_tensorflow/external/local_config_rocm/crosstool/clang/bin/crosstool_wrapper_driver_is_not_gcc:168: SyntaxWarning: invalid escape sequence '\.'
Trying it with
```
diff --git a/third_party/gpus/crosstool/clang/bin/crosstool_wrapper_driver_is_not_gcc.tpl b/third_party/gpus/crosstool/clang/bin/crosstool_wrapper_driver_is_not_gcc.tpl
index 74fafb9b32f..b133db8be94 100755
--- a/third_party/gpus/crosstool/clang/bin/crosstool_wrapper_driver_is_not_gcc.tpl
+++ b/third_party/gpus/crosstool/clang/bin/crosstool_wrapper_driver_is_not_gcc.tpl
@@ -222,7 +222,7 @@ def InvokeNvcc(argv, log=False):
   # Unfortunately, there are other options that have -c prefix too.
   # So allowing only those look like C/C++ files.
   src_files = [f for f in src_files if
-               re.search('\.cpp$|\.cc$|\.c$|\.cxx$|\.C$', f)]
+               re.search('.cpp$|.cc$|.c$|.cxx$|.C$', f)]
   srcs = ' '.join(src_files)
   out = ' -o ' + out_file[0]
   ```

**Comment by hsharsha** [comment]
@jayfurmanek and @zoranjovanovic-ns  Please review this. I have added based on ROCm version compatibility with host compiler as clang 17 in /etc/bazel.bazelrc.

File: tensorflow/tools/tf_sig_build_dockerfiles/setup.rocm.sh [hunk 1]
```diff
@@ -136,6 +136,11 @@ elif [[ "$DISTRO" == "el8" ]]; then
     dnf --enablerepo=extras,epel,elrepo,build_system install -y hipblaslt-devel || true
 fi
 
+function ver { printf "%03d%03d%03d" $(echo "$1" | tr '.' ' '); }
+if [ $(ver "$ROCM_VERSION") -lt $(ver "6.2.0") ]
+then
+  echo "build:rocm --copt=-fclang-abi-compat=17" >> /etc/bazel.bazelrc
```
Comment (jayfurmanek) [inline tensorflow/tools/tf_sig_build_dockerfiles/setup.rocm.sh:142]: This is clever and forward thinking. I likely would have overlooked this.

File: third_party/gpus/rocm_configure.bzl [hunk 2]
```diff
@@ -213,6 +213,8 @@ def _rocm_include_path(repository_ctx, rocm_config, bash_bin):
 
     # Support hcc based off clang 10.0.0 (for ROCm 3.3)
     rocm_toolkit_path = realpath(repository_ctx, rocm_config.rocm_toolkit_path, bash_bin)
+    if int(rocm_config.rocm_version_number) >= 60200:
```
Comment (jayfurmanek) [inline third_party/gpus/rocm_configure.bzl:216]: This was added in https://github.com/ROCm/tensorflow-upstream/pull/2577
Maybe do a pull.

**Review by jayfurmanek — APPROVED** [review]
Looks good. Thanks for working this out.

**Review by zoranjovanovic-ns — APPROVED** [review]

**Pull request merged**